### PR TITLE
Advanced search

### PR DESCRIPTION
**Pull request by q-litzler**

### CODE REVIEW

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/hierarchical/HierarchicalConnectionFilterState.kt [hunk 1]
```diff
@@ -12,30 +10,30 @@ import com.algolia.search.model.filter.Filter
...
-
-    private val filterGroupID = FilterGroupID(hierarchicalGroupName, FilterOperator.And)
+) : ConnectionImpl() {
 
     private val updateSelections: Callback<Filters> = { filters ->
-        viewModel.selections.value = (filters.getFacetFilters(filterGroupID).first().value as Filter.Facet.Value.String).raw.split(viewModel.separator)
+        val hierarchicalFilter = filters.getHierarchicalFilters(viewModel.attribute)
+        val hierarchicalValue = hierarchicalFilter?.filter?.value as Filter.Facet.Value.String
+
+        viewModel.selections.value = hierarchicalValue.raw.split(viewModel.separator)
```
Comment (PLNech) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/hierarchical/HierarchicalConnectionFilterState.kt]: Isn't `hierarchicalValue` nullable here, as it's equal to `hierarchicalFilter?.foo?.bar`?

```suggestion
        viewModel.selections.value = hierarchicalValue?.raw?.split(viewModel.separator)
```

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/state/MutableFiltersImpl.kt [hunk 2]
```diff
@@ -86,7 +95,7 @@ internal data class MutableFiltersImpl(
         tagGroups.filter { it.key !in groupIDs }.forEach { tagGroups.remove(it.key) }
     }
 
-    override fun toString(): String {
-        return "MutableFiltersImpl(facets=$facetGroups, tags=$tagGroups, numerics=$numericGroups)"
+    override fun add(attribute: Attribute, hierarchicalFilter: HierarchicalFilter) {
```
Comment (PLNech) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/state/MutableFiltersImpl.kt]: What about `addOrReplace`, to convey the semantics that writing
```kotlin
add("foo", filterA)
add("foo", filterB)
```
would only leave a single filter for `"foo"`

```suggestion
    override fun addOrReplace(attribute: Attribute, hierarchicalFilter: HierarchicalFilter) {
```

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/hierarchical/HierarchicalConnectionFilterState.kt (hunk 1 above)
Comment (q-litzler) [reply helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/hierarchical/HierarchicalConnectionFilterState.kt]: Fixed.

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/state/MutableFiltersImpl.kt (hunk 2 above)
Comment (q-litzler) [reply helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/filter/state/MutableFiltersImpl.kt]: For now every method is called `add`, so I'd rather keep it that way 🤷‍♂

**Pull request merged**